### PR TITLE
unit: fix umounts manual mounts when there is a failed unit for that …

### PR DESCRIPTION
**Pull request by taskset**
…mount point

We encountered an issue similar to #1741 and #9869. After some analysis,
we may provide the following minimal reproducer.

1, constructing a failed mount point like this:
  mkdir /dev/loop1   /u01
and adding the following incorrect configuration items in /etc/fstab:
  LABEL=/flash1   /u01    ext4   defaults,noatime 0 0

2, Then creating a test loop device:
  dd if=/dev/zero of=floppy.img bs=512 count=2880
  losetup /dev/loop1 floppy.img
  mkfs.ext4 /dev/loop1

3, Mounting this loop device, and we will find it will be immediately
umounted by systemd:
  mount /dev/loop1   /u01
  mount | grep u01

kernel: EXT4-fs (loop1): mounted filesystem without journal. Opts: (null)
kernel: ext4 filesystem being mounted at /u01 supports timestamps until 2038 (0x7fffffff)
systemd[1]: u01.mount: Unit is bound to inactive unit dev-disk-by\x2dlabel-\x5cx2fflash1.device.
systemd[1]: Unmounting /u01...

Finally, we proposed this patch to fix such problems.

### CODE REVIEW

**Comment by poettering** [comment]
The patch isn't right. I mean systemd operates correctly here from what it knows. The thing though is that what it knows is a bit contradicting: it knows info about the backing device from /etc/fstab and info from /proc/self/mountinfo when it is mounted. It currently mixes the two, and thus comes to a BindsTo= dependency based on the fstab info (to a device that is inactive) *and* one to the current block device where it is actually mounted from.

So for the longest time we couldn't fix this, since we couldn't sensibly remove deps once created during runtime. But we have infrastructure for that now. 

The right fix would hence be: 

1. operate as we currently operate when the only data we have about a mount is from /etc/fstab: generate BindsTo= on block device nodes as always
2. when we have both info from /etc/fstab *and* from /proc/self/mountinfo, then remove any BindsTo= deps generated due to rule 1, and instead synthesize BindsTo= on the block device listed in /proc/self/mountinfo
3. when we lose info from /proc/self/mountinfo again (because the device is unmounted), then drop the new BindsTo= again, and switch to the old one.

**Comment by arvidjaar** [comment]
> The right fix would hence be

I miss the reason for such complicated dance. `BindsTo` is needed to stop one active unit when another unit gets deactivated. `BindsTo` on inactive mount unit is useless as unit is already stopped and `BindsTo` on active mount unit is always set based on `/proc/self/mountinfo`. So what are steps 1 and 3 for?

**Comment by w-simon** [comment]
> The patch isn't right. I mean systemd operates correctly here from what it knows. The thing though is that what it knows is a bit contradicting: it knows info about the backing device from /etc/fstab and info from /proc/self/mountinfo when it is mounted. It currently mixes the two, and thus comes to a BindsTo= dependency based on the fstab info (to a device that is inactive) _and_ one to the current block device where it is actually mounted from.
> 
> So for the longest time we couldn't fix this, since we couldn't sensibly remove deps once created during runtime. But we have infrastructure for that now.
> 
> The right fix would hence be:
> 
> 1. operate as we currently operate when the only data we have about a mount is from /etc/fstab: generate BindsTo= on block device nodes as always
> 2. when we have both info from /etc/fstab _and_ from /proc/self/mountinfo, then remove any BindsTo= deps generated due to rule 1, and instead synthesize BindsTo= on the block device listed in /proc/self/mountinfo
> 3. when we lose info from /proc/self/mountinfo again (because the device is unmounted), then drop the new BindsTo= again, and switch to the old one.

Thank you for your comments.

Let's do an experiment:

# systemd-analyze dump
…
-> Unit u01.mount:
…
        BindsTo: dev-disk-by\x2dlabel-\x5cx2fflash1.device (origin-file)
        BindsTo: dev-loop1.device (origin-mountinfo-implicit)

The systemd-analyze dump command displays all the current bindsto information

# journalctl  -b
…
systemd[1]: u01.mount: Job u01.mount/start failed with result 'dependency'.
systemd[1]: dev-disk-by\x2dlabel-\x5cx2fflash1.device: Job dev-disk-by\x2dlabel-\x5cx2fflash1.device/start failed with result 'timeout'.

We can see the error information about the bindsto deps in the log.

In this way, if we modify according to the above suggestions, the information in u01.mount (dev-disk-by\x2dlabel-\x5cx2fflash1.device (origin-file)) will be quietly deleted,
and the systemd-analyze dump will not display complete data. When users analyze problems according to journal logs, they will be confused.


Let's look at suggestions 1, 2 and 3 again. It feels like a priority problem:
When the information in /proc/self/mountinfo exists, it takes precedence; 
when the information in /pro/self/mountinfo does not exist,  fallback to normal /etc/fstab.

Therefore, we may be able to solve this issue with the following simple modifications:
```
diff --git a/src/core/unit.c b/src/core/unit.c
index 35627b3..5ced988 100644
--- a/src/core/unit.c
+++ b/src/core/unit.c
@@ -2016,6 +2016,9 @@ static void unit_check_binds_to(Unit *u) {
         if (unit_active_state(u) != UNIT_ACTIVE)
                 return;

+        if ((u->type == UNIT_MOUNT) && MOUNT(u) && MOUNT(u)->from_proc_self_mountinfo)
+                return;
+
         HASHMAP_FOREACH_KEY(v, other, u->dependencies[UNIT_BINDS_TO], i) {
                 if (other->job)
                         continue;
```
Please take another look, thank you.

**Comment by poettering** [comment]
> I miss the reason for such complicated dance. `BindsTo` is needed to stop one active unit when another unit gets deactivated. `BindsTo` on inactive mount unit is useless as unit is already stopped and `BindsTo` on active mount unit is always set based on `/proc/self/mountinfo`. So what are steps 1 and 3 for?

Step 1+3 is so that we wait for a device to show up before invoking /bin/mount. Don#t forget that BindsTo= also pulls in the device unit into the job queue, so that we have something to wait for before activating the mount.

**Comment by poettering** [comment]
people might add BindsTo= on other units too if they like, for example if they want a mount around only if some service is up or so, the patch you provided would break that logic.

**Comment by w-simon** [comment]
OK, we will improve it according to the step 1+3 soon. 
Thanks.

File: src/core/mount.c [hunk 1]
```diff
@@ -1622,6 +1622,12 @@ static int mount_setup_existing_unit(
 
         MOUNT(u)->from_proc_self_mountinfo = true;
 
+        /* when we have both info from /etc/fstab and from /proc/self/mountinfo, and the mount generated by
+         * fstab is in a failed state then remove any BindsTo= deps generated due to fstab, and instead
+         * snthesize BindsTo= on the block device listed in /proc/self/mountinfo. */
```
Comment (keszybz) [inline src/core/mount.c:1627]: This comment has a bunch of typos: when → When, snthesize. Mounts are not "generated by fstab", at most "generated **from** fstab".

File: src/core/mount.c [hunk 2]
```diff
@@ -1622,6 +1622,12 @@ static int mount_setup_existing_unit(
...
+        if (u->fragment_path && IN_SET(MOUNT(u)->state, MOUNT_DEAD, MOUNT_FAILED))
```
Comment (keszybz) [inline src/core/mount.c:1628]: I don't understand why the condition is like this here. It doesn't match what @poettering wrote in https://github.com/systemd/systemd/pull/15389#issuecomment-629514514, and it doesn't seem right: the removal of dependencies generated based on /etc/fstab shouldn't be conditional. So I think this here should be just
```
    unit_remove_dependencies(u, UNIT_DEPENDENCY_FILE);
```
(no `if`),

File: src/core/mount.c [hunk 3]
```diff
@@ -1956,6 +1964,12 @@ static int mount_process_proc_self_mountinfo(Manager *m) {
                                 break;
                         }
 
+                        /* When we have both info from /etc/fstab and /proc/self/mountinfo, we discard the dep info
+                         * from /etc/fstab and add the dep info from /proc/self/mountinfo; when we lose the info
+                         * from /proc/self/mountinfo again (because the device is unmounted), the we need to add the
```
Comment (keszybz) [inline src/core/mount.c:1969]: the we → then we ?

**Review by keszybz — COMMENTED** [review]
I'm not sure if I grok the patch...

**Comment by poettering** [comment]
What systemd version did you see the actual problem on?

**Comment by w-simon** [comment]
Thanks, we checked the latest systemd and found that this problem has been fixed.
So let's close this pr.

**Comment by jficz** [comment]
what version is this supposed to be fixed in? It's still present for me in systemd 239 (Centos8). Maybe close/update https://github.com/systemd/systemd/issues/1741

**Pull request closed without merging**